#' A simulated phylogenetic tree with 10 tips and 9 internal nodes
#'
#' A phlo object randomly created by using the function \code{\link[ape]{rtree}}
#'
#' @format A phylo object with 10 tips and 9 internal nodes:
#' \describe{
#'   \item{Tip labels}{t1, t2, ..., t10.}
#'   \item{Node labels}{Node_11, Node_12, ..., Node_19}
#' }
#'
#'
"tinyTree"


#' A simulated phylogenetic tree with 50 tips and 49 internal nodes
#'
#' A phlo object randomly created by using the function \code{\link[ape]{rtree}}
#'
#' @format A phylo object with 10 tips and 9 internal nodes:
#' \describe{
#'   \item{Tip labels}{t1, t2, ..., t50.}
#'   \item{Node labels}{NULL}
#' }
"exTree"


#' A simulated phylogenetic tree with 800 tips and 799 internal nodes
#'
#' A phlo object randomly created by using the function \code{\link[ape]{rtree}}
#'
#' @format A phylo object with 10 tips and 9 internal nodes:
#' \describe{
#'   \item{Tip labels}{t1, t2, ..., t800.}
#'   \item{Node labels}{NULL}
#' }
"bigTree"

#' AML-sim data
#'
#' A semi-simulated count table generated by computationally "spiking in" small percentages of AML blast cells into healthy BMMCs (created by Lukas Weber)
#'
#' @format A data frame of cell counts collected from healthy and cytogenetically normal (CN) samples:
#' \describe{
#'   \item{cluster}{The label of cluster}
#'   \item{p_vals}{The original p value}
#'   \item{p_adj}{The adjusted p value}
#'   \item{n_cells}{The number of cells in the cluster}
#'   \item{n_spikein}{The number of spike-in cells}
#'   \item{prop_spikein}{The proportion of the spike-in cell}
#'   \item{true}{whether the cluster is truly differential abundant}
#'   \item{healthy_H1}{The counts gained from the first healthy sample}
#'   \item{healthy_H2}{The counts gained from the second healthy sample}
#'   \item{healthy_H3}{The counts gained from the third healthy sample}
#'   \item{healthy_H4}{The counts gained from the forth healthy sample}
#'   \item{healthy_H5}{The counts gained from the fifth healthy sample}
#'   \item{CN_H1}{The counts gained from the first CN sample}
#'   \item{CN_H2}{The counts gained from the second CN sample}
#'   \item{CN_H3}{The counts gained from the third CN sample}
#'   \item{CN_H4}{The counts gained from the forth CN sample}
#'   \item{CN_H5}{The counts gained from the fifth CN sample}
#' }
"res_table"

#' the count table of AML-sim data
#'
#' A semi-simulated count table generated by computationally "spiking in" small percentages of AML blast cells into healthy BMMCs (created by Lukas Weber)
#'
#' @format A data frame of cell counts collected from healthy and cytogenetically normal (CN) samples:
#' \describe{
#'   \item{healthy_H1}{The counts gained from the first healthy sample}
#'   \item{healthy_H2}{The counts gained from the second healthy sample}
#'   \item{healthy_H3}{The counts gained from the third healthy sample}
#'   \item{healthy_H4}{The counts gained from the forth healthy sample}
#'   \item{healthy_H5}{The counts gained from the fifth healthy sample}
#'   \item{CN_H1}{The counts gained from the first CN sample}
#'   \item{CN_H2}{The counts gained from the second CN sample}
#'   \item{CN_H3}{The counts gained from the third CN sample}
#'   \item{CN_H4}{The counts gained from the forth CN sample}
#'   \item{CN_H5}{The counts gained from the fifth CN sample}
#' }
"simCount"


#' The true signal in AML-sim data
#'
#' This data provides the differential abundance information of clusters in AML-sim data
#'
#' @format A data frame of two columns:
#' \describe{
#'   \item{cluster}{The label of cluster}
#'    \item{true}{whether the cluster is truly differential abundant}
#' }
"simTruth"

#' The hiearchical tree of the cells
#'
#' A hclust object which records the hiearchical structure of the cells.
#'
#' @format A hclust object with 900 leaves. Each leaf is a cell cluster.
"simTree"

#' OTU count table from 16S sequencing of the throat microbiome samples.
#'
#' @format More details could be found in \code{\link[GUniFrac]{throat.otu.tab}}
"throat.otu.tab"

#' UPGMA tree of the OTUs from 16S sequencing of the throat microbiome samples.
#'
#' @format More details could be found in \code{\link[GUniFrac]{throat.tree}}
"throat.tree"

